### PR TITLE
Codegen fixed size for static sympy values

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99362
* #99491

cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/actions/runs/4724868015

Deberta dynamic + 10% ish on local, hoping for some nice speedups from this.

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1448,20 +1447,24 @@ def codegen_static_numels(self, code):
...
+                postfix = "# dynamic_shapes=False" if not dynamo_config.dynamic_shapes else ""
+                simplified_tree_numel = V.graph.sizevars.simplify(tree.numel)
+                if isinstance(simplified_tree_numel, sympy.Integer):
                     code.writeline(
-                        f"{tree.prefix}numel = {V.graph.sizevars.size_hint(tree.numel)}"
-                    )
-                elif not dynamo_config.dynamic_shapes:
-                    code.writeline(
-                        f"{tree.prefix}numel = {V.graph.sizevars.size_hint(tree.numel)}  # dynamic_shapes=False"
+                        f"{tree.prefix}numel = {int(simplified_tree_numel)} {postfix}"
```
Comment (ezyang) [inline torch/_inductor/codegen/triton.py:1454]: This is more of a pre-existing condition thing, but how does the downstream code know if you wrote out a static numel or not? Is downstream testing the same simplify to Integer stuff?

Comment (voznesenskym) [reply torch/_inductor/codegen/triton.py:1454]: Good question - what this does is stomp on the passed-in inputs by hardcoding them.

Truncated top of kernel looks like:

```
@reduction(
    size_hints=[4096, 1024],
    reduction_hint=ReductionHint.DEFAULT,
    filename=__file__,
    meta={'signature': {0: '*i64', 1: '*fp32', 2: '*i64', 3: '*fp32', 4: '*fp32', 5: '*fp32', 6: '*fp32', 7: 'i32', 8: 'i32'}, 'device': 0, 'constants': {}, 'mutated_arg_names': [], 'configs': [instance_descriptor(divisible_by_16=(), equal_to_1=())]}
)
@triton.jit
def KERNEL_NAME(in_ptr0, in_ptr1, in_ptr2, in_ptr3, in_ptr4, in_ptr5, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
```

And this code writes

```
    xnumel = 4096
    rnumel = 768
```

Up after the signature, before the code block. As its hardcoded, it becomes a better signal to triton on how to unroll and do some static indexing. So, it's not so much that downstream knows that its a static numel, as that you just plop a constant into the kernel.

Comment (ezyang) [reply torch/_inductor/codegen/triton.py:1454]: Would be a nice comment

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Can you perf test this

**Comment by voznesenskym** [comment]
> Can you perf test this

Run kicked off above :)

**Comment by voznesenskym** [comment]
Inference speedup: 3% TB, 4% HF, 2% Timm 
Training speedup: 2% TB, N/A HF,  2% Timm

Not insane, but not bad.

Added a small check and kicked off again, I think theres another percent or two. 

Stacked another potentially good PR on top, but that one is oh-so-bad atm.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1445,23 +1444,40 @@ def codegen_kernel(self, name=None):
...
-                # we generate a static RBLOCK for persistent reduction
-                hint = V.graph.sizevars.size_hint(tree.numel)
-                hint = next_power_of_2(hint)
-                code.writeline(f"RBLOCK: tl.constexpr = {hint}")
+                simplified_tree_numel = V.graph.sizevars.simplify(tree.numel)
+                if dynamo_config.dynamic_shapes:
+                    if isinstance(simplified_tree_numel, (sympy.Integer, int)):
+                        val = int(simplified_tree_numel)
+                    else:
+                        continue
```
Comment (ezyang) [inline torch/_inductor/codegen/triton.py:1476]: Why continue?

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -1445,23 +1444,40 @@ def codegen_kernel(self, name=None):
...
+                        f"{tree.prefix}numel = {int(simplified_tree_numel)} {postfix}"
                     )
 
             if tree.prefix == "r" and self.persistent_reduction:
-                # we generate a static RBLOCK for persistent reduction
-                hint = V.graph.sizevars.size_hint(tree.numel)
-                hint = next_power_of_2(hint)
-                code.writeline(f"RBLOCK: tl.constexpr = {hint}")
+                simplified_tree_numel = V.graph.sizevars.simplify(tree.numel)
+                if dynamo_config.dynamic_shapes:
```
Comment (ezyang) [inline torch/_inductor/codegen/triton.py:1472]: Why do you want to `dynamic_shapes` gate this?

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_inductor/codegen/triton.py:1478]: we're in a loop and theres nothing left to do with this - it's dynamic.

**Pull request closed without merging**